profiling/perfetto-connector/libperfetto-connector.cpp

```diff
@@ -0,0 +1,182 @@
+  printf("KokkosP: Perfetto Connector (sequence is %d, version: %lu)\n",
+         loadSeq, interfaceVer);
+  printf("-----------------------------------------------------------\n");
+
+  perfetto::TracingInitArgs args{};
+  perfetto::TraceConfig cfg{};
+  perfetto::protos::gen::TrackEventConfig track_event_cfg{};
+
+  // default 10 MB shared memory and 1 GB buffer
+  auto shmem_size_hint = get_env("KOKKOSP_PERFETTO_SHMEM_SIZE_HINT_KB", 10000);
+  auto buffer_size = get_env("KOKKOSP_PERFETTO_BUFFER_SIZE_KB", 1000 * 1000);
+  auto _fill_policy_name = get_env("KOKKOSP_PERFETTO_FILL_POLICY", "discard");
+
+  if (std::getenv("KOKKOSP_PERFETTO_INPROCESS_BACKEND") &&
+      std::stoi(std::getenv("KOKKOSP_PERFETTO_INPROCESS_BACKEND")) != 0)
+    args.backends |= perfetto::kInProcessBackend;
+  if (std::getenv("KOKKOSP_PERFETTO_SYSTEM_BACKEND") &&
+      std::stoi(std::getenv("KOKKOSP_PERFETTO_SYSTEM_BACKEND")) != 0)
+    args.backends |= perfetto::kInProcessBackend;
+
+  if (args.backends == 0) args.backends |= perfetto::kInProcessBackend;
+
+  auto _fill_policy =
+      _fill_policy_name == "discard"
+          ? perfetto::protos::gen::TraceConfig_BufferConfig_FillPolicy_DISCARD
+          : perfetto::protos::gen::
+                TraceConfig_BufferConfig_FillPolicy_RING_BUFFER;
+
+  auto* buffer_config = cfg.add_buffers();
+  buffer_config->set_size_kb(buffer_size);
+  buffer_config->set_fill_policy(_fill_policy);
+
+  auto* ds_cfg = cfg.add_data_sources()->mutable_config();
+  ds_cfg->set_name("track_event");  // this MUST be track_event
+  ds_cfg->set_track_event_config_raw(track_event_cfg.SerializeAsString());
+
+  args.shmem_size_hint_kb = shmem_size_hint;
+
+  perfetto::Tracing::Initialize(args);
+  perfetto::TrackEvent::Register();
+
+  tracing_session = perfetto::Tracing::NewTrace();
+  tracing_session->Setup(cfg);
+  tracing_session->StartBlocking();
+}
+
+void kokkosp_finalize_library() {
+  printf("-----------------------------------------------------------\n");
+  printf("KokkosP: Finalizing Perfetto Connector... ");
+  fflush(stdout);
+
+  perfetto::TrackEvent::Flush();
+  tracing_session->FlushBlocking();
+  tracing_session->StopBlocking();
+
+  printf("Reading trace data... ");
+  fflush(stdout);
+
+  std::vector<char> _trace_data{tracing_session->ReadTraceBlocking()};
+
+  if (!_trace_data.empty()) {
+    auto _fname =
+        get_env("KOKKOSP_PERFETTO_OUTPUT_FILE", "kokkosp-perfetto.pftrace");
+
+    std::ofstream _ofs{_fname};
+    if (_ofs) {
+      printf("Writing '%s'... ", _fname.c_str());
+      fflush(stdout);
+      _ofs.write(&_trace_data[0], _trace_data.size());
+    } else
+      fprintf(stderr, "Error opening %s!\n", _fname.c_str());
+  }
+
+  printf("Done\n");
+  printf("-----------------------------------------------------------\n");
+  fflush(stdout);
+}
+
+void kokkosp_begin_parallel_for(const char* name, const uint32_t, uint64_t*) {
+  TRACE_EVENT_BEGIN("kokkos.kernels", perfetto::StaticString(name));
```

Comment:
Not distinguishing `parallel_for` from `parallel_reduce` and `parallel_scan`?
I don't necessary have an issue with it.  Just wondering if that is by design.

Reply:
Nope not by design, just inherited this from David's. It's a fair point though and easy to fix